### PR TITLE
Fixed #27301 -- Prevented exceptions that fail unpickling from crashing the parallel test runner.

### PR DESCRIPTION
**Pull request by amw**
Until now we've only been checking whether the error can be pickled, but some errors get through that and then fail during unpickling in the main process causing the Pool management code to blow up.

One example of error like that is botocore.exceptions.ClientError which pickles as "ClientError(message)", but its initializer is actually ClientError(response, operation_name). This causes the parallel test runner to fail with:
```
multiprocessing/connection.py", line 251, in recv
    return ForkingPickler.loads(buf.getbuffer())
TypeError: __init__() missing 1 required positional argument: 'operation_name'
```
CC @aaugustin @cjerdonek

### CODE REVIEW

**Comment by timgraham** [comment]
A Trac ticket and a regression test is also needed.

**Comment by amw** [comment]
I was trying to find a place where you test behavior of parallel test runner and came up empty. And there's certainly no place where you test this particular `check_pickleable` function. If there was already a test that ensured errors that fail `pickle.dumps` are reported then I would gladly replicate it for errors that fail `pickle.loads`. Otherwise the required additions to the testing infrastructure are too much for me. This is a small enhancement to GH PR #4761 (Trac #20461).

I've briefly scanned contributors guide and it said that trivial issues are fine as Github PR without Trac issue. This is two lines of code deep in test-runner edge case so I thought I will be fine, but I can create a Trac issue if you insist.

**Comment by timgraham** [comment]
Yes, please, on creating a ticket. CONTRIBUTING.rst says "non-trivial pull requests (anything more than fixing a typo) without Trac tickets will be closed". If there's another place in the docs we could update, let me know. Maybe 42dcceba614081966dfb5f236ae1b6fe78f31683 gives some ideas about how to test it.

**Comment by amw** [comment]
Trac-ed in #27301. The problem with testing this is that the current behavior raises error that stops the entire test suite. This is another thing I wanted to tackle in a separate PR. Maybe then it will be easier to test it. I've described the problem in Trac.

**Comment by cjerdonek** [comment]
> I was trying to find a place where you test behavior of parallel test runner and came up empty. And there's certainly no place where you test this particular check_pickleable function.

I would suggest defining a helper function for the following line in your PR (both because doing this makes it testable and because you repeat this line twice in your PR):

    pickle.loads(pickle.dumps(err))
    
And then add tests for that function.

In addition, I would follow @timgraham's suggestion and try looking at commit  42dcceb of mine. It contains `test_add_failing_subtests()`, which provides an example of an "end-to-end" test of a failing parallel test -- without interrupting the test run.

**Comment by amw** [comment]
Working on a separate method with a unit test. Coming soon.

That `test_add_failing_subtests` is a simple test for a "proper" failure: the runner collects results, reports them to stderr on the main thread and finishes cleanly.

Pickle errors on the other hand cause the runner to abort in child process and just blow up: `multiprocessing.Pool` ends with `RemoteTraceback`. A clean test case like this would be possible once the second part of my Trac ticket is accepted and we start encapsulating the pickle errors in a container and sending them back to the main thread.

**Comment by amw** [comment]
Added unit test and rebased on current master.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by amw** [comment]
@timgraham Fixed errors on python 2.7 and rebased on master.

File: tests/test_runner/test_parallel.py [hunk 1]
```diff
@@ -10,6 +10,13 @@
     tblib = None
 
 
+class ErrorThatFailsUnpickling(Exception):
```
Comment (timgraham) [inline tests/test_runner/test_parallel.py]: Add a comment to explain why this fails unpickling?

File: tests/test_runner/test_parallel.py [hunk 2]
```diff
@@ -10,6 +10,13 @@
...
+    def __init__(self, arg1, arg2):
+        self.arg1 = arg1
+        self.arg2 = arg2
+        super(ErrorThatFailsUnpickling, self).__init__("Message with %s and %s")
```
Comment (timgraham) [inline tests/test_runner/test_parallel.py]: Usually we're using single quotes for strings unless the string contains single quotes.

File: tests/test_runner/test_parallel.py [hunk 3]
```diff
@@ -44,6 +51,16 @@ def dummy_test(self):
...
 class RemoteTestResultTest(SimpleTestCase):
 
+    def test_pickle_errors_detection(self):
+        picklable_error = RuntimeError("This is fine")
+        not_unpicklable_error = ErrorThatFailsUnpickling("one arg", "another arg")
+
+        result = RemoteTestResult()
+        result._raise_when_not_picklable(picklable_error)
+
+        with self.assertRaisesMessage(TypeError, "argument"):
```
Comment (timgraham) [inline tests/test_runner/test_parallel.py]: Including a bit more of the message might help readability to understand exactly what `TypeError` is raised..

**Comment by amw** [comment]
I updated the commit according to your comments.

**Comment by amw** [comment]
Sorry about my commits not being grouped as a single review. I didn't had the option with the page that I opened.

**Comment by amw** [comment]
That's what I was saying all along. There is no test that checked the original `pickle.dumps` error handling, had it been there I would add another one that also checked for `pickle.loads`. All I could do is add a unit test for the new method, as you requested.

File: tests/test_runner/test_parallel.py [hunk 4]
```diff
@@ -10,6 +10,17 @@
     tblib = None
 
 
+class ExceptionThatFailsUnpickling(Exception):
+    """
+    This class fails unpickling with an error about incorrect arguments passed
```
Comment (cjerdonek) [inline tests/test_runner/test_parallel.py]: I would clarify, "After pickling, this class" or "This class fails unpickling (after pickling) with".

File: tests/test_runner/test_parallel.py [hunk 5]
```diff
@@ -10,6 +10,17 @@
...
+    to __init__().
+    """
+    def __init__(self, arg1, arg2):
+        self.arg1 = arg1
+        self.arg2 = arg2
+        super(ExceptionThatFailsUnpickling, self).__init__('Message with %s and %s')
```
Comment (cjerdonek) [inline tests/test_runner/test_parallel.py]: Does the format string `'Message with %s and %s'` serve a purpose here? If so, a comment might be good. It doesn't seem like it is necessary.

File: tests/test_runner/test_parallel.py [hunk 6]
```diff
@@ -44,6 +55,17 @@ def dummy_test(self):
...
 
+    def test_pickle_errors_detection(self):
+        picklable_error = RuntimeError('This is fine')
+        not_unpicklable_error = ExceptionThatFailsUnpickling('one arg', 'another arg')
+
+        result = RemoteTestResult()
+        result._confirm_picklable(picklable_error)
+
+        # __init__() missing 1 required positional argument (different on PY2)
+        with self.assertRaisesMessage(TypeError, 'argument'):
```
Comment (cjerdonek) [inline tests/test_runner/test_parallel.py]: I would be in favor of including the full exception text for the two cases with `if PY2`, etc. It's more cryptic than it needs to be. And when switching to Django 2, the `PY2` block can simply be deleted rather than having to remember to fill in the fuller message.

Comment (timgraham) [inline tests/test_runner/test_parallel.py]: We could do that. I had planned to simply fill in the Python 3 message in a few months when Python 2 support is dropped in master.
